admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/instance/AgentInstanceEndpoints.kt

```diff
@@ -61,92 +63,80 @@ object Agents {
-            completeSendingClassMetadataRoute()
-        }
-    }
-
-    private fun Route.registerAgentInstanceRoute() {
-        put<Agents> {
-            val agentConfig = call.decompressAndReceive<AgentConfig>()
-            val agentInfo: AgentInfo = withContext(Dispatchers.IO) {
-                agentManager.attach(agentConfig)
-            }
-            processPluginData(agentInfo, InitInfo())
-            call.addDrillInternalHeader()
-            call.respond(HttpStatusCode.OK)
-        }
-    }
-
-    private fun Route.sendCoverageRoute() {
-        post<Agents.Coverage> { params ->
-            handleAgentRequest(params.agentId, params.buildVersion) { agentInfo ->
-                val data = call.decompressAndReceive<CoverageData>()
-                processPluginData(agentInfo, data.toCoverDataPart())
+fun Routing.agentInstanceRoutes() {
+    val app by closestDI().instance<Application>()
+    val agentManager by closestDI().instance<AgentManager>()
+    val pd by closestDI().instance<PluginDispatcher>()
+    authenticate("api-key") {
+        withRole(Role.USER, Role.ADMIN) {
+            put<Agents> {
+                val agentConfig = call.decompressAndReceive<AgentConfig>()
+                val agentInfo: AgentInfo = withContext(Dispatchers.IO) {
+                    agentManager.attach(agentConfig)
+                }
+                processPluginData(pd, agentInfo, InitInfo())
                 call.addDrillInternalHeader()
+                call.respond(HttpStatusCode.OK)
             }
-        }
-    }
 
-    private fun Route.sendClassMetadataRoute() {
-        post<Agents.ClassMetadata> { params ->
-            handleAgentRequest(params.agentId, params.buildVersion) { agentInfo ->
-                val data = call.decompressAndReceive<ClassMetadata>()
-                processPluginData(agentInfo, data.toInitDataPart())
-                call.addDrillInternalHeader()
+            post<Agents.Coverage> { params ->
+                handleAgentRequest(agentManager, params.agentId, params.buildVersion) { agentInfo ->
+                    val data = call.decompressAndReceive<CoverageData>()
+                    processPluginData(pd, agentInfo, data.toCoverDataPart())
+                    call.addDrillInternalHeader()
+                }
             }
-        }
-    }
 
-    private fun Route.completeSendingClassMetadataRoute() {
-        post<Agents.ClassMetadataComplete> { params ->
-            handleAgentRequest(params.agentId, params.buildVersion) { agentInfo ->
-                processPluginData(agentInfo, Initialized())
-                call.addDrillInternalHeader()
+            post<Agents.ClassMetadata> { params ->
+                handleAgentRequest(agentManager, params.agentId, params.buildVersion) { agentInfo ->
+                    val data = call.decompressAndReceive<ClassMetadata>()
+                    processPluginData(pd, agentInfo, data.toInitDataPart())
+                    call.addDrillInternalHeader()
```

Comment:
It is better to place the `drill-internal` header inside the `handleAgentRequest` function, and send it to the agent in all cases, not only in case of success.

Reply:
done